src/rgw/rgw_kms.cc

```diff
@@ -0,0 +1,275 @@
+  if (vault_url.empty()) {
+    ldout(cct, 0) << "ERROR: Vault URL is not set" << dendl;
+    return -EINVAL;
+  }
+
+  std::string secret_url = vault_url + std::string(key_id);
+  RGWHTTPTransceiver secret_req(cct, "GET", secret_url, secret_bl);
+  secret_req.append_header("X-Vault-Token", vault_token);
+  res = secret_req.process(null_yield);
+  if (res <= 0) {
+    ldout(cct, 0) << "ERROR: Request to Vault failed with error " << res << dendl;
+    return res;
+  }
+
+  if (secret_req.get_http_status() ==
+      RGWHTTPTransceiver::HTTP_STATUS_UNAUTHORIZED) {
+    ldout(cct, 0) << "ERROR: Vault request failed authorization" << dendl;
+    return -EACCES;
+  }
+
+  ldout(cct, 20) << "Request to Vault returned " << res << " and HTTP status "
+    << secret_req.get_http_status() << dendl;
+  return res;
+}
+
+int get_actual_key_from_vault(CephContext *cct,
+                              boost::string_view key_id,
+                              std::string& actual_key)
+{
+  int res = 0;
+  std::string auth;
+  bufferlist secret_bl;
+
+  auth = cct->_conf->rgw_crypt_s3_kms_vault_auth;
+  ldout(cct, 20) << "Vault auhentication method " << auth << dendl;
+
+  if ("token" == auth) {
+    res = request_key_from_vault_with_token(cct, key_id, &secret_bl);
+  } else if ("agent" == auth) {
+    res = request_key_from_vault_with_agent(cct, key_id, &secret_bl);
+  } else {
+    ldout(cct, 0) << "ERROR: Unsupported authentication method " << auth << dendl;
+    return -EINVAL;
+  }
+
+  if (res < 0) {
+    return res;
+  }
+
+  JSONParser parser;
+  if (!parser.parse(secret_bl.c_str(), secret_bl.length())) {
+    ldout(cct, 0) << "ERROR: Failed to parse JSON response from Vault" << dendl;
+    return -EINVAL;
+  }
+
+  // TODO: handle possible missing structures in the JSON response
+  JSONObj *data_obj = parser.find_obj("data");
+  string secret = from_base64(data_obj->find_obj("data")->find_obj("key")->get_data());
+
+  // TODO: remove this log entry
+  ldout(cct, 20) << "Vault secret length: " << secret.length() << dendl;
+
+  actual_key.assign(secret.c_str(), secret.length());
+  memset(secret_bl.c_str(), 0, secret_bl.length());
+
+  return res;
+}
+
+int get_actual_key_from_kms(CephContext *cct,
+                            boost::string_view key_id,
+                            boost::string_view key_selector,
+                            std::string& actual_key)
+{
+  std::string kms_backend;
+
+  kms_backend = cct->_conf->rgw_crypt_s3_kms_backend;
+  ldout(cct, 20) << "Getting KMS encryption key for key=" << key_id << dendl;
+  ldout(cct, 20) << "SSE-KMS backend is " << kms_backend << dendl;
+
+  if (RGW_SSE_KMS_BACKEND_BARBICAN == kms_backend)
+    return get_actual_key_from_barbican(cct, key_id, actual_key);
+
+  if (RGW_SSE_KMS_BACKEND_VAULT == kms_backend)
+    return get_actual_key_from_vault(cct, key_id, actual_key);
+
+  if (RGW_SSE_KMS_BACKEND_LOCAL != kms_backend)
+    ldout(cct, 10) << "Unknown SSE-KMS backend, reverting to local" << dendl;
```

Comment:
error message should include the name of rgw_crypt_s3_kms_backend. and since this is pointing out a configuration issue that should be fixed, i think it's appropriate to print at debug level 1

Reply:
Good point!